spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarPropertiesMapper.java

```diff
@@ -52,11 +59,53 @@ final class PulsarPropertiesMapper {
 	void customizeClientBuilder(ClientBuilder clientBuilder, PulsarConnectionDetails connectionDetails) {
 		PulsarProperties.Client properties = this.properties.getClient();
 		PropertyMapper map = PropertyMapper.get().alwaysApplyingWhenNonNull();
-		map.from(connectionDetails::getBrokerUrl).to(clientBuilder::serviceUrl);
 		map.from(properties::getConnectionTimeout).to(timeoutProperty(clientBuilder::connectionTimeout));
 		map.from(properties::getOperationTimeout).to(timeoutProperty(clientBuilder::operationTimeout));
 		map.from(properties::getLookupTimeout).to(timeoutProperty(clientBuilder::lookupTimeout));
 		customizeAuthentication(clientBuilder::authentication, properties.getAuthentication());
+		customizeServiceUrlProviderBuilder(clientBuilder::serviceUrlProvider, properties, connectionDetails);
+		if (properties.getFailover().getBackupClusters().isEmpty()) {
```

Comment:
Because this condition controls an "either-or", I think it would be more readable to not do the check in the `customizeServiceUrlProviderBuilder` and instead do an `if/else` here.